rmw_fastrtps_shared_cpp/src/rmw_take.cpp

```diff
@@ -75,14 +75,16 @@ _take(
   data.data = ros_message;
   data.impl = info->type_support_impl_;
   if (info->subscriber_->takeNextData(&data, &sinfo)) {
-    info->listener_->data_taken(info->subscriber_);
+    info->listener_->updateDataCount(info->subscriber_);
 
     if (eprosima::fastrtps::rtps::ALIVE == sinfo.sampleKind) {
       if (message_info) {
         _assign_message_info(identifier, message_info, &sinfo);
       }
       *taken = true;
     }
+  } else {
+    info->listener_->updateDataCount(info->subscriber_);
```

Comment:
could you elaborate why adding this will solve the problem and it should be added even when data is not taken.

Reply:
rmw_fastrtps_cpp use a variable `data_` in SubListener to check if there is a topic data by `onNewDataMessage` which is called by FastDDS if there is a new change (`mp_history->received_change`)
but if FastDDS removed the change, it will not call a function to notify an interface of SubListener.

About why or in what kinds of cases will FastDDS remove the change, I need more time to investigate it.

Reply:
> About why or in what kinds of cases will FastDDS remove the change, I need more time to investigate it.

Maybe this is hard to investigate. It is better to consult eProsmia team.

Reply:
@Barry-Xu-2018

Yes. This PR maybe is just a workaround, not a solution.
It is better to call `eProsima` team in.